### ISSUE TITLE
 calabash 0.20.4 ports 37265 and 27753

### ISSUE THREAD

**Issue by Johncalabash** [calabash/calabash-ios#1304]
We are having a peculiar issue  ..
     Xcode 8
     iOS 10
     iOS  calabash 0.20.4 

     runloop 2.4.1 

    it starts the simulator and install the app .. 

    with the port 27753 and then later fails  indicating connection error with 37265 port .

Below I have included the complete log ...
so how do we make it work without any issues.

[36mSHELL: xcrun simctl list devices --json [0m
[35mDEBUG: Simulator instruction set 'x86_64' is compatible with '["x86_64"]' [0m
[35mDEBUG: Simulator relaunch required: simulator is not running. [0m
[35mDEBUG: Simulator relaunch not required: no running apps [0m
[35mDEBUG: Sending 'KILL' to UserEventAgent process '60' [0m
[35mDEBUG: Cannot kill process '60' with 'KILL'; not a child of this process [0m
[35mDEBUG: Waiting for UserEventAgent with pid '60' to terminate [0m
[35mDEBUG: Waited for 0.527668 seconds for UserEventAgent with pid '60' to terminate [0m
[35mDEBUG: Sending 'KILL' to UserEventAgent process '64432' [0m
[35mDEBUG: Waiting for UserEventAgent with pid '64432' to terminate [0m
[35mDEBUG: Waited for 0.108626 seconds for UserEventAgent with pid '64432' to terminate [0m
[35mDEBUG: Sending 'KILL' to mobileassetd process '76' [0m
[35mDEBUG: Cannot kill process '76' with 'KILL'; not a child of this process [0m
[35mDEBUG: Waiting for mobileassetd with pid '76' to terminate [0m
[35mDEBUG: Waited for 0.517287 seconds for mobileassetd with pid '76' to terminate [0m
[35mDEBUG: Launching #<Simulator: iPhone 6 (10.0) 744F1E1C-549F-4A8E-9269-034E552AEAE4 x86_64> with: [0m
[36mSHELL: xcrun open -g -a /Applications/Xcode.app/Contents/Developer/Applications/Simulator.app --args -CurrentDeviceUDID 744F1E1C-549F-4A8E-9269-034E552AEAE4 LAUNCHED_BY_RUN_LOOP [0m
[35mDEBUG: Waited for 0.178863 seconds for 'Simulator' to start. [0m
[35mDEBUG: Waiting for simulator to stabilize with timeout: 240 seconds [0m
[35mDEBUG: Timed out after 240 seconds waiting for simulator to stabilize [0m
[35mDEBUG: Took 274.006764 seconds to launch the simulator [0m
[36mSHELL: xcrun simctl install 744F1E1C-549F-4A8E-9269-034E552AEAE4 /Users/go/Documents/TUI/Syed/Appfiles/NL.app [0m
[35mDEBUG: Waited for 2.0e-06 seconds for app to install [0m
[35mDEBUG: HTTP: get http://127.0.0.1:27753/1.0/health {:retries=>1, :timeout=>0.5} [0m
[35mDEBUG: HTTP: get http://127.0.0.1:27753/1.0/health {:retries=>1, :timeout=>0.5} [0m
[35mDEBUG: Waiting for DeviceAgent to launch... [0m
[36mSHELL: cd /Users/go/.rvm/rubies/ruby-2.2.0/lib/ruby/gems/2.2.0/gems/run_loop-2.4.1/lib/run_loop/device_agent [0m
[36mSHELL: ditto -xk Frameworks.zip . [0m
[36mSHELL: ditto -xk DeviceAgent-Runner.app.zip . [0m

Retrying.. Errno::ECONNREFUSED: (Connection refused - Connection refused - connect(2) for "127.0.0.1" port 37265 (127.0.0.1:37265))
Retrying.. Errno::ECONNREFUSED: (Connection refused - Connection refused - connect(2) for "127.0.0.1" port 37265 (127.0.0.1:37265))

**Comment by jmoody** [comment]
The DeviceAgent server is on 27753.

The LPServer, which is embedded in your application, runs on port 37265.

The error you are reporting means that your application is not linked with the LPServer (calabash.framework).

**Comment by Johncalabash** [comment]
Hi Jmoody,
Thanks for the reply. I am looking into this issue to ensure that the application is linked with the LPServer and I will let you know the results .

Best Regards
John

**Comment by Johncalabash** [comment]
Hi Jmoody,

 Xcode 8.3.2
 iOS 10.3
 iOS  calabash 0.20.4 
 runloop 2.4.1 
Locally it works fine but doesn't work on jenkins server ....Our jenkins server is in the cloud and it interacts with the slave OSX machines and runs the automations in the slave OS X machines. This setup  used to work fine before the  Xcode 8.3.2 and ios 10.3 and device agent runner app introduction.
it starts the simulator and install the app ..
with the port 27753 and then later fails  indicating connection error with 37265 port  in Jenkins server ... though it works fine in local machine.  can you please let us know the issue.

**Comment by jmoody** [comment]
> Jenkins

We cannot provide support for executing in CI environments - they are too hard for us to debug.

As a sanity check I can tell you that we use Jenkins and Travis to test Calabash.

**Comment by Johncalabash** [comment]
Hi Jmoody,

before with  Xcode 7 ,  ios 9 and ios calabash 0.15,  runloop 2.2.1 
it used work fine on jenkins and generate reports. we have made no changes to the jenkins setup.
we only upgraded to xcode 8.3.2 ;iOS 10.3; iOS calabash 0.20.4
runloop 2.4.1 and the deviceAgent runner that we are getting these issues.
So is there something missing with calabash 0.20.4 or runloop 2.4.1 ?? why is it causing issues but had no issues earlier ???

Best Regards
John

**Comment by JoeSSS** [comment]
We have the same setup that you described and everything works fine on
Jenkins. Are you sure that you updated the server version inside the app?
Not just the gem

**Comment by Johncalabash** [comment]
Hi JoeSSS and Jmoody 
Please find the outputs related to versions -- 0.204 
it is the same for calabash client, calabash server and calabash linked to the app.
There is no change to the Jenkins setup as it is the same and did work perfectly fine until this upgrade. 

1)Calabash ios client  is updated ..  0.20.4
~ $: calabash-ios version
0.20.4

2) Calabash version linked to the app ..
~ $: xcrun strings /Users/tthsa1/Desktop/testlegoios/Appfiles/NL.app/NL | grep -E 'CALABASH VERSION'
CALABASH VERSION: 0.20.4

3)Calabash ios server embedded in the app  is also at 0.20.4
~ $: curl http://localhost:37265/version
{"device_family":"iPhone","outcome":"SUCCESS","server_port":37265,"simulator_device":"","simulator":"CoreSimulator 375.21 - Device: iPhone 6 - Runtime: iOS 10.3 (14E269) - DeviceType: iPhone 6","app_name":"TUI Nederland","app_version":"1","screen_dimensions":{"native_scale":2,"sample":1,"height":1334,"width":750,"scale":2},"git":{"revision":"dc15597","remote_origin":"git@github.com:calabash\/calabash-ios-server.git","bran~ $: ~ 

Please let me know what are the next steps .
Kind Regards
Syed

**Comment by JoeSSS** [comment]
Hi @Johncalabash ,
Looks like you're using old iOS Simulator version on your Jenkins node. I'm not sure what is the problem with iOS 10.0 but for me it looks like the latest calabash versions have problems with iOS less than 10,3. 
For our Jenkins nodes I removed all the simulators that are less than iOS 10.3. As alternative you can also specify iOS version explicitly (e.g. ```export Device="iPhone 7"  
export OS="10.3"  
export DEVICE_TARGET="$Device ($OS)" ```

**Comment by jmoody** [comment]
> OS

Calabash does not respond to this variable.

http://calabashapi.xamarin.com/ios/file.ENVIRONMENT_VARIABLES.html

**Comment by Johncalabash** [comment]
Hi @JoeSSS ,

Thanks I will look into this and get back to you ; if this is indeed the issue or if it is not .
Please let me go over it during the weekend and I will let you know the result on monday .

Best Regards
John

**Comment by JoeSSS** [comment]
@jmoody yeah, I use $OS inside DEVICE_TARGET. To, have more control on CI between iOS version and phone model

**Comment by Johncalabash** [comment]
Hi @jmoody   @JoeSSS ;   

Thanks, we will try out these steps in CI nodes and keep you informed about the result .  
Best Regards
John

**Comment by AnnawanDev** [comment]
Was there ever a solution?  I'm experiencing the same issue.  If I remote into a node, I can run Calabash from a shell window.  It works just fine.  If I SSH into the same node, or use Jenkins, then I get,

`Errno::ECONNREFUSED: (Connection refused - Connection refused - connect(2) for "127.0.0.1" port 37265 (127.0.0.1:37265))`

So problem seems related to the shell and whether you're executing remotely vs from the machine itself.
